AssemblerLib/VectorMatrixAssembler.h

```diff
@@ -49,15 +57,41 @@ class VectorMatrixAssembler
     {
         assert(_data_pos.size() > id);
 
-        LocalToGlobalIndexMap::RowColumnIndices const& indices = _data_pos[id];
+        std::vector<GlobalIndexType> indices;
+
+        // Local matrices and vectors will always be ordered by component,
+        // no matter what the order of the global matrix is.
+        for (unsigned c=0; c<_data_pos.getNumComponents(); ++c)
+        {
+            auto const& idcs = _data_pos(id, c).rows;
+            indices.reserve(indices.size() + idcs.size());
```

Comment:
Just related the negative ID for PETSc, which you mentioned to put it in the next PR in your reply to the next comment.